patroni/postgresql/rewind.py

```diff
@@ -333,10 +346,13 @@ def execute(self, leader):
         #   running a checkpoint or
         #   waiting until Patroni on the master will expose checkpoint_after_promote=True
         checkpoint_status = leader.checkpoint_after_promote if isinstance(leader, Leader) else None
-        if checkpoint_status is None:  # master still runs the old Patroni
-            leader_status = self._postgresql.checkpoint(self._conn_kwargs(leader, self._postgresql.config.superuser))
-            if leader_status:
-                return logger.warning('Can not use %s for rewind: %s', leader.name, leader_status)
+        if checkpoint_status is None:  # we are the standby-cluster leader or master still runs the old Patroni
+            timeline_status = self.check_leader_has_run_checkpoint(r)
+            if timeline_status:  # we checked the leader's timeline in pg_control, and it is outdated
+                leader_status = self._postgresql.checkpoint(
+                        self._conn_kwargs(leader, self._postgresql.config.superuser))
+                if leader_status:  # we tried to run a checkpoint on the remote leader, but it failed
+                    return logger.warning('Can not use %s for rewind: %s', leader.name, leader_status)
```

Comment:
The `check_leader_has_run_checkpoint()` doesn't return anything. Also it has the Postgres version-specific query, i.e. it will fail with 9.6 and older (I know that it is EOL, but the reality shows that many people still run it).

With Postgres 10 and older, we don't even need to call the `check_leader_has_run_checkpoint()`, because the superuser access must be allowed anyway for pg_rewind to work and we can immediately do a CHECKPOINT.

So the code should do something like:
```suggestion
            # The superuser credential match with rewind_credentials if the last one are not provided or we run 10 or older
            if self._postgresql.config.superuser == self._postgresql.config.rewind_credentials:
                leader_status = self._postgresql.checkpoint(
                        self._conn_kwargs(leader, self._postgresql.config.superuser))
            else:  # We run 11+ and have a dedicated pg_rewind user
                leader_status = self.check_leader_has_run_checkpoint(r)

            if leader_status:  # we tried to run a checkpoint on the remote leader, but it failed
                return logger.warning('Can not use %s for rewind: %s', leader.name, leader_status)
```

Of course, the `check_leader_has_run_checkpoint()` should return a string explaining what went wrong.

Reply:
OK, I did that now, many thanks for the suggestions